garbage_collector.rst

```diff
@@ -0,0 +1,460 @@
+count field initialized to the reference count (``gc_ref`` in the figures)
+of that object when the algorithm starts. This is because the algorithm needs
+to modify the reference count to do the computations and in this way the
+interpreter will not modify the real reference count field. 
+
+.. figure:: images/python-cyclic-gc-1-new-page.png
+
+The GC then iterates over all containers in the first list and decrements by one the
+``gc_ref`` field of any other object that container is referencing.  Doing
+this makes use of the ``tp_traverse`` slot in the container class (implemented
+using the C API or inherited by a superclass) to know what objects are referenced by
+each container. After all the objects have been scanned, only the objects that have
+references from outside the “objects to scan” list will have ``gc_ref > 0``.
+
+.. figure:: images/python-cyclic-gc-2-new-page.png
+
+Notice that having ``gc_refs == 0`` does not imply that the object is unreachable.
+This is because another object that is reachable from the outside (``gc_refs > 0``)
+can still have references to it. For instance, the ``link_2`` object in our example
+ended having ``gc_refs == 0`` but is referenced still by the ``link_1`` object that
+is reachable from the outside. To obtain the set of objects that are really
+unreachable, the garbage collector scans again the container objects using the
+``tp_traverse`` slot with a different traverse function that marks objects with
+``gc_refs == 0`` as "tentatively unreachable" and then moves them to the
+tentatively unreachable list. The following image depicts the state of the lists in a
+moment when the GC processed the ``link 3`` and ``link 4`` objects but has not
+processed ``link 1`` and ``link 2`` yet.
+
+.. figure:: images/python-cyclic-gc-3-new-page.png
+
+Then the GC scans the next ``link 1`` object. Because its has ``gc_refs == 1``
+the gc does not do anything special because it knows it has to be reachable (and is
+already in what will become the reachable list):
+
+.. figure:: images/python-cyclic-gc-4-new-page.png
+
+When the GC encounters an object which is reachable (``gc_refs > 0``), it traverses
+its references using the ``tp_traverse`` slot to find all the objects that are
+reachable from it, moving them to the end of the list of reachable objects (where
+they started originally) and setting its ``gc_refs`` field to 1. This is what happens
+to ``link 2`` and ``link 3`` below as they are reachable from ``link 1``.  From the
+state in the previous image and after examining the objects referred to by ``link1``
+the GC knows that ``link 3`` is reachable after all, so it is moved back to the
+original list and its ``gc_refs`` field is set to one so if the GC visits it again, it
+does know that is reachable. To avoid visiting a object twice, the GC marks all
+objects that are not visited yet with and once an object is processed is unmarked so
+the GC does not process it twice.
```

Comment:
I couldn't follow the sentence just above, and don't know what it's intending to say so can't suggest an improvement :wink:.

Reply:
Whoops! That sentence was never finished. I fixed it in 5ca46f7